src/Microsoft.FeatureManagement/FeatureFilters/Recurrence/RecurrenceEvaluator.cs

```diff
@@ -0,0 +1,1117 @@
+                RemainingDaysOfTheWeek(DateTime.Today.DayOfWeek, firstDayOfWeek));
+
+            DateTime prev = DateTime.MinValue;
+
+            TimeSpan minGap = TimeSpan.FromDays(DayNumberOfWeek);
+
+            for (int i = 0; i < DayNumberOfWeek; i++)
+            {
+                if (daysOfWeek.Any(day =>
+                    day == date.DayOfWeek))
+                {
+                    if (prev == DateTime.MinValue)
+                    {
+                        //
+                        // Find a occurrence for the first time
+                        prev = date;
+                    }
+                    else
+                    {
+                        TimeSpan gap = date - prev;
+
+                        if (gap < minGap)
+                        {
+                            minGap = gap;
+                        }
+
+                        prev = date;
+                    }
+                }
+
+                date += TimeSpan.FromDays(1);
+            }
+
+            //
+            // It may across weeks. Check the next week if the interval is one week.
+            if (interval == 1)
+            {
+                for (int i = 0; i < DayNumberOfWeek; i++)
+                {
+                    if (daysOfWeek.Any(day =>
+                        day == date.DayOfWeek))
+                    {
+                        TimeSpan gap = date - prev;
+
+                        if (gap < minGap)
+                        {
+                            minGap = gap;
+                        }
+
+                        //
+                        // Only check the first occurrence in the next week
+                        break;
+                    }
+
+                    date += TimeSpan.FromDays(1);
+                }
+            }
+
+            return minGap >= duration;
+        }
+
+        private static int RemainingDaysOfTheWeek(DayOfWeek dayOfWeek, DayOfWeek firstDayOfWeek)
+        {
+            int remainingDays = (int) dayOfWeek - (int) firstDayOfWeek;
+
+            if (remainingDays < 0)
+            {
+                return -remainingDays;
+            }
+            else
+            {
+                return DayNumberOfWeek - remainingDays;
+            }
+        }
+
+        /// <summary>
+        /// Find the nth day of week in the month of the date time.
+        /// </summary>
+        /// <param name="dateTime">A date time.</param>
+        /// <param name="index">The index of the day of week in the month.</param>
+        /// <param name="dayOfWeek">The day of week.</param>
+        /// <returns>The data time of the nth day of week in the month.</returns>
+        private static DateTime NthDayOfWeekInTheMonth(DateTime dateTime, WeekIndex index, DayOfWeek dayOfWeek)
+        {
+            var date = new DateTime(dateTime.Year, dateTime.Month, 1);
+
+            //
+            // Find the first provided day of week in the month
+            while (date.DayOfWeek != dayOfWeek)
```

Comment:
Just wondering why you decided to loop over the days instead of some kind of subtraction between `dayOfWeek` and `date.DayOfWeek`?

Reply:
Oh, good suggestion. Perhaps I was not in my clear mind when I was coding at that time. Yes, indeed, subtraction should be used here, or the above DaysOfTheWeek helper function could be utilized.